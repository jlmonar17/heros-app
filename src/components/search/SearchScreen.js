import { useLocation } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import { HeroCard } from "../heros/HeroCard";
import queryString from "query-string";
import { getHerosByName } from "../../selectors/getHerosByName";
import { useMemo } from "react";

export const SearchScreen = ({ history }) => {
    // Get prop location from props of the component
    const location = useLocation();

    // Detructuring to get only query value that I need
    const { q = "" } = queryString.parse(location.search);

    const { formValues, handleInputChange } = useForm({
        search: q,
    });

    const { search } = formValues;

    // Get heroes array only when "q" value changes, otherwise, heroes array will be memorized
    const herosFiltered = useMemo(() => getHerosByName(q), [q]);

    const handleSearch = (e) => {
        e.preventDefault();

        // Added query string to current url
        history.push(`?q=${search}`);
    };

    return (
        <>
            <h1>Search Screen</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h5>Search Form</h5>
                    <hr />

                    <form onSubmit={handleSearch}>
                        <input
                            type="text"
                            name="search"
                            placeholder="Your search"
                            className="form-control"
                            autoComplete="off"
                            value={search}
                            onChange={handleInputChange}
                        />

                        <button
                            type="submit"
                            className="btn w-100 btn-outline-primary mt-3"
                        >
                            Search
                        </button>
                    </form>
                </div>

                <div className="col-7">
                    <h4>Results</h4>
                    <hr />

                    {q === "" && (
                        <div className="alert alert-info">Search a hero...</div>
                    )}

                    {q !== "" && herosFiltered.length === 0 && (
                        <div className="alert alert-danger">
                            No heros found with {q}
                        </div>
                    )}

                    {herosFiltered.map((hero) => (
                        <HeroCard key={hero.id} {...hero} />
                    ))}
                </div>
            </div>
        </>
    );
};
