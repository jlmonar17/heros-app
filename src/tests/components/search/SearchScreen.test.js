import { mount } from "enzyme";
import { MemoryRouter, Route } from "react-router-dom";
import { SearchScreen } from "../../../components/search/SearchScreen";

describe("Tests for <SearchScreen />", () => {
    test("Should render correctly with values by default", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find(".alert-info").text().trim()).toBe(
            "Search a hero..."
        );
    });

    test("Should show Batman character and input with value of queryString", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batman"]}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find("input").prop("value")).toBe("batman");
        expect(wrapper).toMatchSnapshot();
    });

    test("Should show error if hero doesn't exist", () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={["/search?q=batmanaaaa"]}>
                <Route path="/search" component={SearchScreen} />
            </MemoryRouter>
        );

        expect(wrapper.find(".alert-danger").exists()).toBe(true);
        expect(wrapper.find(".alert-danger").text().trim()).toBe(
            "No heros found with batmanaaaa"
        );
        expect(wrapper).toMatchSnapshot();
    });

    test("Should call history.push", () => {
        const history = {
            push: jest.fn(),
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={["/search"]}>
                <Route
                    path="/search"
                    component={(props) => <SearchScreen history={history} />}
                />
            </MemoryRouter>
        );

        wrapper.find("input").simulate("change", {
            target: {
                name: "search",
                value: "batman",
            },
        });

        wrapper.find("form").simulate("submit", { preventDefault: () => {} });

        expect(history.push).toHaveBeenCalled();
        expect(history.push).toHaveBeenCalledWith("?q=batman");
    });
});
