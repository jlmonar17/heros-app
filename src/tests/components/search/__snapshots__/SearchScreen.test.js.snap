// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for <SearchScreen /> Should render correctly with values by default 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h5>
        Search Form
      </h5>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Your search"
          type="text"
          value=""
        />
        <button
          className="btn w-100 btn-outline-primary mt-3"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero...
      </div>
    </div>
  </div>,
]
`;

exports[`Tests for <SearchScreen /> Should show Batman character and input with value of queryString 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h5>
        Search Form
      </h5>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Your search"
          type="text"
          value="batman"
        />
        <button
          className="btn w-100 btn-outline-primary mt-3"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="col"
      >
        <div
          className="card h-100"
          style={
            Object {
              "maxWidth": 540,
            }
          }
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="./assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              className="col-md-8"
            >
              <h5
                className="card-title"
              >
                Batman
              </h5>
              <p
                className="card-text"
              >
                Bruce Wayne
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                  Detective Comics #27
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                More...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Tests for <SearchScreen /> Should show error if hero doesn't exist 1`] = `
Array [
  <h1>
    Search Screen
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h5>
        Search Form
      </h5>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Your search"
          type="text"
          value="batmanaaaa"
        />
        <button
          className="btn w-100 btn-outline-primary mt-3"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No heros found with 
        batmanaaaa
      </div>
    </div>
  </div>,
]
`;
